presto-parquet/src/main/java/io/prestosql/parquet/reader/ShortDecimalColumnReader.java

```diff
@@ -47,54 +52,104 @@
+                throw new ParquetDecodingException(format("Unsupported Presto column type (%s) for Parquet column (%s)", prestoType, columnDescriptor));
+            }
+
+            long value;
+
             // When decimals are encoded with primitive types Parquet stores unscaled values
             if (columnDescriptor.getPrimitiveType().getPrimitiveTypeName() == INT32) {
-                decimalValue = valuesReader.readInteger();
+                value = valuesReader.readInteger();
             }
             else if (columnDescriptor.getPrimitiveType().getPrimitiveTypeName() == INT64) {
-                decimalValue = valuesReader.readLong();
+                value = valuesReader.readLong();
             }
             else {
-                decimalValue = getShortDecimalValue(valuesReader.readBytes().getBytes());
+                value = getShortDecimalValue(valuesReader.readBytes().getBytes());
             }
 
-            if (isShortDecimal(prestoType)) {
-                validateDecimal((DecimalType) prestoType);
-                // TODO: validate that value fits Presto decimal precision
-                prestoType.writeLong(blockBuilder, decimalValue);
-            }
-            else if (isLongDecimal(prestoType)) {
-                validateDecimal((DecimalType) prestoType);
-                // TODO: validate that value fits Presto decimal precision
-                prestoType.writeSlice(blockBuilder, unscaledDecimal(decimalValue));
+            DecimalType prestoDecimalType;
+
+            if (prestoType instanceof DecimalType) {
+                prestoDecimalType = (DecimalType) prestoType;
             }
-            else if (prestoType.equals(TINYINT) || prestoType.equals(SMALLINT) || prestoType.equals(INTEGER) || prestoType.equals(BIGINT)) {
+            else {
+                // for numeric types convert decimal to unscaled long
+                prestoDecimalType = DecimalType.createDecimalType(MAX_SHORT_PRECISION, 0);
+
                 if (parquetDecimalType.getScale() != 0) {
-                    throw new ParquetDecodingException(format(
-                            "Parquet decimal column type with non-zero scale (%s) cannot be converted to Presto %s column type",
-                            parquetDecimalType.getScale(),
-                            prestoType));
+                    throw new PrestoException(NOT_SUPPORTED, format("Unsupported Presto column type (%s) for Parquet column (%s)", prestoType, columnDescriptor));
                 }
-                prestoType.writeLong(blockBuilder, decimalValue);
             }
-            else {
-                throw new ParquetDecodingException(format("Unsupported Presto column type (%s) for Parquet column (%s)", prestoType, columnDescriptor));
+
+            long rescale = longTenToNth(Math.abs(prestoDecimalType.getScale() - parquetDecimalType.getScale()));
+
+            if (isShortDecimal(prestoDecimalType)) {
+                long scaledValue = shortToShortCast(
```

Comment:
"scaledValue" is a wrong name, because it's "unscaled value in presto decimal type", right?

"convertedValue"?